runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -629,27 +631,40 @@ protected void rollbackStreamUnsafe(ISMRStream stream, long rollbackVersion) {
             return;
         }
 
-        List<SMREntry> entries = stream.current();
+        boolean isHole = false;
+        List<SMREntry> entries = null;
+        try {
+            entries = stream.current();
+        } catch (HoleInStreamException ex) {
+            isHole = true;
+        }
 
-        while (!entries.isEmpty()) {
-            if (entries.stream().allMatch(SMREntry::isUndoable)) {
-                // start from the end, process one at a time
-                ListIterator<SMREntry> it = entries.listIterator(entries.size());
-                while (it.hasPrevious()) {
-                    applyUndoRecordUnsafe(it.previous(), stream);
-                }
-            } else {
-                Optional<SMREntry> entry = entries.stream().findFirst();
-                if (log.isTraceEnabled()) {
-                    log.trace("rollbackStreamUnsafe: one or more stream entries in address @{} are not undoable. " +
-                                    "Undoable entries: {}/{}", stream.pos(),
-                            (int) entries.stream().filter(SMREntry::isUndoable).count(),
-                            entries.size());
+        while (isHole || !entries.isEmpty()) {
+            if (!isHole) {
+                if (entries.stream().allMatch(SMREntry::isUndoable)) {
+                    // start from the end, process one at a time
+                    ListIterator<SMREntry> it = entries.listIterator(entries.size());
+                    while (it.hasPrevious()) {
+                        applyUndoRecordUnsafe(it.previous(), stream);
+                    }
+                } else {
+                    Optional<SMREntry> entry = entries.stream().findFirst();
+                    if (log.isTraceEnabled()) {
+                        log.trace("rollbackStreamUnsafe: one or more stream entries in address @{} are not undoable. " +
+                                        "Undoable entries: {}/{}", stream.pos(),
+                                (int) entries.stream().filter(SMREntry::isUndoable).count(),
+                                entries.size());
+                    }
+                    throw new NoRollbackException(entry, stream.pos(), rollbackVersion);
```

Comment:
I think its cleaner to just have an SMR that is a whole which is always undoable. When applyUndoRecordUnsafe is called, it just results in a no-op.